### ISSUE TITLE
3.2 sync_support.c:do_folders() can pass uninitialized pointer to sync_rename_list_free

### ISSUE THREAD

**Issue by dilyanpalauzov** [cyrusimap/cyrus-imapd#3209]
Compiling imapd-3.2 with `gcc-10.2 -flto -O3` reveals in imap/sync_support.c:do_folders():
```c
    struct sync_rename_list *rename_folders;
    struct sync_reserve_list *reserve_list;
   ...
    if (r) goto bail;
    rename_folders = sync_rename_list_create();
    reserve_list = sync_reserve_list_create(SYNC_MSGID_LIST_HASH_SIZE);
 bail:
    sync_rename_list_free(&rename_folders);
    sync_reserve_list_free(&reserve_list);
```
gcc prints: rename_folders/reserve_list may be uninitialized when passed to sync_rename_list_free/sync_reserve_list_free()

**Comment by elliefm** [comment]
Thanks, this is now fixed on the 3.2 branch.  Master seems unaffected

**Comment by dilyanpalauzov** [comment]
A better fix would have been to replace the `goto bail;` with `return r;`.
